CONTRIBUTING.md

```diff
@@ -121,3 +122,115 @@ run it before submitting your PR. It runs `gofmt -s` (simplify) and `golint`.
+A running Collector typically contains data that is received but not yet exported further
+(e.g. is stored in the queues and other processors). Crashing or exiting Collector
+process will result in losing this data since typically the receiver have
+already acknowledged the receipt for this data and the senders of the data will
+not send that data again. This recommendation may change in the future if we
+implement persistent queues in the Collector.
+
+### Bad Input Handling
+
+Do not crash on bad input in receivers or elsewhere in the pipeline.
+[Crash-only software](https://en.wikipedia.org/wiki/Crash-only_software)
+is valid in certain cases however this is not a correct approach for Collector (except
+during startup, see above). The reason is that many senders from which Collector
+receives data have built-in automatic retries of the _same_ data if no
+acknowledgment is received from the Collector. If you crash on bad input
+chances are high that after the Collector is restarted it will see the same
+data in the input and will crash again. This will likely result in infinite
+crashing loop if you have automatic retries in place.
+
+### Error Handling and Retries
+
+Be rigorous in error handling. Don't ignore errors. Think carefully about each
+error and decide if it is a fatal problem or a transient problem that may go away
+when retried. For fatal errors that the users of Collector need to be aware of
+log them or have an internal metric that counts occurrences. For transient
+errors come up with a retrying strategy and implement it. Typically you will
+want to implement retries with some sort of exponential back-off strategy. For
+connection or sending retries use jitter for back-off intervals to avoid overwhelming
+your destination when network is restored or the destination is recovered.
+[Exponential Backoff](https://github.com/cenkalti/backoff) is a good library that
+provides all this functionality.
+
+### Logging
+
+Log your component startup and shutdown, including successful outcomes (but don't
+overdo it, keep the number of success message to a minimum).
+This can help to understand the context of failures if they occur elsewhere after
+your code is successfully executed.
+
+Use logging carefully for events that can happen frequently to avoid flooding
+the logs. Avoid outputting logs per a received or processed data item since this can
+amount to very large number of log entries (Collector is designed to process
+many thousands of spans and metrics per second). For such high-frequency events
+instead of logging consider adding an internal metric and increment it when
+the event happens.
+
+Make log message human readable and also include data that is needed for easier
+understanding of what happened and in what context.
+
+### Resource Usage
+
+Limit usage of CPU, RAM or other resources that the code can use. Do not let the code
+to consume a resource in an uncontrolled manner. For example if you have a queue
```

Comment:
Do not let the code to consume a resource in an uncontrolled manner.
=>
Do not write code that consumes resources in an uncontrolled manner.

Reply:
Fixed.